# Feature Engineering

After completing data cleaning and exploratory data analysis, feature engineering becomes the crucial step that transforms raw attributes into meaningful inputs for modeling. By encoding categorical variables, scaling numeric features, and deriving clinically relevant labels, we ensure that the dataset is standardized, and optimized for machine learning. This process enhances model accuracy, prevents bias from uneven feature ranges, and captures the underlying patterns that drive predictive performance.


**1. Import Libraries**

In [0]:
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.ml.feature import VectorAssembler, MinMaxScaler
from pyspark.ml.feature import StringIndexer
from pyspark.ml import Pipeline
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

**2. Load Data**

In [0]:
# load data to dataframe and display 
df = spark.table('workspace.default.diabetes_data_explored')
display(df.limit(5))

cholesterol_total,stabilized_glucose,hdl_cholesterol,chol_hdl_ratio,location,age,gender,body_frame,systolic_bp_1,diastolic_bp_1,minutes_post_meal,BMI,waist_hip_ratio,hba1c_category
203.0,82.0,56.0,3.6,Buckingham,46,female,medium,118.0,59.0,720,22.130944261888523,0.7631578947368421,Normal
165.0,97.0,24.0,6.9,Buckingham,29,female,large,112.0,68.0,360,37.41920003371257,0.9583333333333334,Normal
228.0,92.0,37.0,6.2,Buckingham,58,female,large,190.0,92.0,180,48.37024068028249,0.8596491228070176,Normal
78.0,93.0,12.0,6.5,Buckingham,67,male,large,110.0,50.0,480,18.637828409539644,0.868421052631579,Normal
249.0,90.0,28.0,8.9,Buckingham,64,male,medium,138.0,80.0,300,27.824746566448155,1.0731707317073171,Diabetic


In [0]:
display(df.count())

387

**3. Split Data**

In [0]:
# Split the data
splits = df.randomSplit([0.7, 0.3])
train = splits[0]
test = splits[1]
print ("Training Rows:", train.count(), " Testing Rows:", test.count())

Training Rows: 272  Testing Rows: 115


**4. Encode Categorical Columns**

Encode the `hba1c_category`,  `location`, `gender` and `body_frame` categorical column values as numeric.

In [0]:
# Define indexers for each categorical column
indexer_label = StringIndexer(inputCol="hba1c_category", outputCol="label")
indexer_location = StringIndexer(inputCol="location", outputCol="locationIdx")
indexer_gender = StringIndexer(inputCol="gender", outputCol="genderIdx")
indexer_body_frame = StringIndexer(inputCol="body_frame", outputCol="bodyFrameIdx")

# Chain them together in a pipeline
pipeline = Pipeline(stages=[indexer_label, indexer_location, indexer_gender, indexer_body_frame])

# Fit and transform the training data
indexedData = (pipeline.fit(train).transform(train).drop("location", "gender", "body_frame", "minutes_post_meal"))

display(indexedData.limit(5))

cholesterol_total,stabilized_glucose,hdl_cholesterol,chol_hdl_ratio,age,systolic_bp_1,diastolic_bp_1,BMI,waist_hip_ratio,hba1c_category,label,locationIdx,genderIdx,bodyFrameIdx
78.0,93.0,12.0,6.5,67,110.0,50.0,18.637828409539644,0.868421052631579,Normal,0.0,1.0,1.0,2.0
118.0,95.0,39.0,3.0,47,140.0,76.0,21.112667908929566,0.8333333333333334,Normal,0.0,0.0,0.0,1.0
129.0,110.0,42.0,3.1,56,140.0,75.0,19.387037970569732,0.8947368421052632,Normal,0.0,1.0,1.0,1.0
132.0,83.0,40.0,3.3,28,136.0,86.0,34.210753975141174,0.7884615384615384,Normal,0.0,0.0,0.0,0.0
132.0,99.0,34.0,3.9,21,112.0,62.0,28.122760245520492,0.9069767441860465,Normal,0.0,1.0,0.0,2.0


In [0]:
# show hba1c_category encoded mapping
display(indexedData.select("hba1c_category", "label").distinct())

hba1c_category,label
Diabetic,1.0
Normal,0.0


In [0]:
#drop hba1c_category column
indexedData = indexedData.drop("hba1c_category")

**5. Normalize Numeric Features**

Numeric values often exist on different ranges. During model training, the absolute units of measurement matter less than the relative differences between observations. If one feature has much larger values, it can dominate the learning process and distort the model’s predictions. To prevent this imbalance, numeric features are typically scaled to a common range — for example, mapping values to decimals between 0.0 and 1.0.

In [0]:
# Create a vector column containing all numeric features
numericFeatures = [c for c in indexedData.columns if c not in ["label", "locationIdx", "genderIdx", "bodyFrameIdx"]]
numericColVector = VectorAssembler(inputCols=numericFeatures, outputCol="numericFeatures")
vectorizedData = numericColVector.transform(indexedData)


In [0]:
# Use a MinMax scaler to normalize the numeric values in the vector
minMax = MinMaxScaler(inputCol = numericColVector.getOutputCol(), outputCol="normalizedFeatures")
scaledData = minMax.fit(vectorizedData).transform(vectorizedData)

In [0]:
# Display the data with numeric feature vectors (before and after scaling) 
display(scaledData.select("numericFeatures", "normalizedFeatures").limit(5))

numericFeatures,normalizedFeatures
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""78.0"",""93.0"",""12.0"",""6.5"",""67.0"",""110.0"",""50.0"",""18.637828409539644"",""0.868421052631579""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.0"",""0.12302839116719243"",""0.0"",""0.446601951298699"",""0.6666666666666666"",""0.125"",""0.02702702702702703"",""0.06621910241429305"",""0.4047442550037067""]}"
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""118.0"",""95.0"",""39.0"",""3.0"",""47.0"",""140.0"",""76.0"",""21.112667908929566"",""0.8333333333333334""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.12269938650306748"",""0.12933753943217666"",""0.25"",""0.10679612055003723"",""0.38888888888888884"",""0.3125"",""0.3783783783783784"",""0.12842912693762312"",""0.3286384976525824""]}"
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""129.0"",""110.0"",""42.0"",""3.0999999046325684"",""56.0"",""140.0"",""75.0"",""19.387037970569732"",""0.8947368421052632""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.15644171779141106"",""0.17665615141955834"",""0.2777777777777778"",""0.11650484931245347"",""0.5138888888888888"",""0.3125"",""0.3648648648648649"",""0.0850519783144412"",""0.46182357301704985""]}"
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""132.0"",""83.0"",""40.0"",""3.299999952316284"",""28.0"",""136.0"",""86.0"",""34.210753975141174"",""0.7884615384615384""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.16564417177914112"",""0.0914826498422713"",""0.25925925925925924"",""0.1359223299847212"",""0.125"",""0.28750000000000003"",""0.5135135135135136"",""0.4576756318477709"",""0.23131094257854828""]}"
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""132.0"",""99.0"",""34.0"",""3.9000000953674316"",""21.0"",""112.0"",""62.0"",""28.122760245520492"",""0.9069767441860465""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.16564417177914112"",""0.14195583596214512"",""0.2037037037037037"",""0.19417477200152436"",""0.027777777777777776"",""0.1375"",""0.1891891891891892"",""0.30464177275874926"",""0.48837209302325585""]}"


In [0]:
#Prepare feature and labels for training
featVect = VectorAssembler(inputCols=["locationIdx", "genderIdx", "bodyFrameIdx", "normalizedFeatures"], outputCol="featuresVector")
preppedData = featVect.transform(scaledData)[col("featuresVector").alias("features"), col("label").alias("label")]
display(preppedData.limit(5))


features,label
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""1.0"",""2.0"",""0.0"",""0.12302839116719243"",""0.0"",""0.446601951298699"",""0.6666666666666666"",""0.125"",""0.02702702702702703"",""0.06621910241429305"",""0.4047442550037067""]}",0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.0"",""0.0"",""1.0"",""0.12269938650306748"",""0.12933753943217666"",""0.25"",""0.10679612055003723"",""0.38888888888888884"",""0.3125"",""0.3783783783783784"",""0.12842912693762312"",""0.3286384976525824""]}",0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""1.0"",""1.0"",""0.15644171779141106"",""0.17665615141955834"",""0.2777777777777778"",""0.11650484931245347"",""0.5138888888888888"",""0.3125"",""0.3648648648648649"",""0.0850519783144412"",""0.46182357301704985""]}",0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.0"",""0.0"",""0.0"",""0.16564417177914112"",""0.0914826498422713"",""0.25925925925925924"",""0.1359223299847212"",""0.125"",""0.28750000000000003"",""0.5135135135135136"",""0.4576756318477709"",""0.23131094257854828""]}",0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""0.0"",""2.0"",""0.16564417177914112"",""0.14195583596214512"",""0.2037037037037037"",""0.19417477200152436"",""0.027777777777777776"",""0.1375"",""0.1891891891891892"",""0.30464177275874926"",""0.48837209302325585""]}",0.0



> **Note**: The dataset has been fully prepared for machine learning.  
>- Categorical fields were indexed.  
>- Numeric features were normalized.  
>- A diabetes status label was derived from `hba1c_percent`.  
>- All predictors were assembled into a single feature vector.  

> We now have a clean structure of **features** and **labels**, ready to move into the machine learning phase for training and evaluation.


**6. Model Training**

In [0]:
#Train the machine learning model
lr = LogisticRegression(labelCol="label", featuresCol="features", maxIter=10, regParam=0.3)
model = lr.fit(preppedData)
print ("Model trained!")


Model trained!


In [0]:

# Prepare the test data for testing the model
# Fit and transform the test data
indexedTestData = (pipeline.fit(test).transform(test).drop("location", "gender", "body_frame", "hba1c_category"))
vectorizedTestData = numericColVector.transform(indexedTestData)
scaledTestData = minMax.fit(vectorizedTestData).transform(vectorizedTestData)
preppedTestData = featVect.transform(scaledTestData)[col("featuresVector").alias("features"), col("label").alias("label")]


In [0]:
# Get predictions
prediction = model.transform(preppedTestData)
predicted = prediction.select("features", "probability", col("prediction").astype("Int"), col("label").alias("trueLabel"))
display(predicted.limit(5))

features,probability,prediction,trueLabel
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.0"",""0.0"",""0.0"",""0.0"",""0.10089020771513353"",""0.20430107526881722"",""0.18055555371590595"",""0.2328767123287671"",""0.11111111111111112"",""0.33333333333333337"",""0.24317381702405433"",""0.5569135802469136""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.9398649928886776"",""0.06013500711132236""]}",0,0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""1.0"",""0.0"",""0.03409090909090909"",""0.5192878338278932"",""0.0"",""0.5277778182500685"",""0.5616438356164384"",""0.11111111111111112"",""0.15151515151515152"",""0.3946867816118515"",""0.7322997416020671""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.6915650877275387"",""0.30843491227246134""]}",0,1.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""0.0"",""0.0"",""0.06818181818181818"",""0.15727002967359052"",""0.12903225806451613"",""0.30555557027275215"",""0.0821917808219178"",""0.4666666666666667"",""0.30303030303030304"",""0.7665975103734441"",""0.2876436781609194""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.8838531621620216"",""0.11614683783797841""]}",0,0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""0.0"",""2.0"",""0.07386363636363637"",""0.11869436201780416"",""0.10752688172043012"",""0.3472222314204701"",""0.136986301369863"",""0.2"",""0.045454545454545456"",""0.09516192393626163"",""0.0""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.9408115734384279"",""0.05918842656157208""]}",0,0.0
"{""type"":""1"",""size"":null,""indices"":null,""values"":[""1.0"",""0.0"",""1.0"",""0.07954545454545454"",""0.09792284866468844"",""0.2903225806451613"",""0.16666667770456414"",""0.0410958904109589"",""0.05555555555555556"",""0.4090909090909091"",""0.2519975309488701"",""0.4604166666666667""]}","{""type"":""1"",""size"":null,""indices"":null,""values"":[""0.9446785016225897"",""0.055321498377410294""]}",0,0.0


**7. Model Evaluation**

In [0]:
# Get evaluation metrics for the classification model
evaluator = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction")

# Simple accuracy
accuracy = evaluator.evaluate(prediction, {evaluator.metricName:"accuracy"})
print("Accuracy:", accuracy)

# Individual class metrics
labels = [0,1]
print("\nIndividual class metrics:")
for label in sorted(labels):
   print ("Class %s" % (label))

   # Precision
   precision = evaluator.evaluate(prediction, {evaluator.metricLabel:label,
                                               evaluator.metricName:"precisionByLabel"})
   print("\tPrecision:", precision)

   # Recall
   recall = evaluator.evaluate(prediction, {evaluator.metricLabel:label,
                                            evaluator.metricName:"recallByLabel"})
   print("\tRecall:", recall)

   # F1 score
   f1 = evaluator.evaluate(prediction, {evaluator.metricLabel:label,
                                        evaluator.metricName:"fMeasureByLabel"})
   print("\tF1 Score:", f1)

# Weighted (overall) metrics
overallPrecision = evaluator.evaluate(prediction, {evaluator.metricName:"weightedPrecision"})
print("Overall Precision:", overallPrecision)
overallRecall = evaluator.evaluate(prediction, {evaluator.metricName:"weightedRecall"})
print("Overall Recall:", overallRecall)
overallF1 = evaluator.evaluate(prediction, {evaluator.metricName:"weightedFMeasure"})
print("Overall F1 Score:", overallF1)




Accuracy: 0.8869565217391304

Individual class metrics:
Class 0
	Precision: 0.8807339449541285
	Recall: 1.0
	F1 Score: 0.9365853658536586
Class 1
	Precision: 1.0
	Recall: 0.3157894736842105
	F1 Score: 0.4799999999999999
Overall Precision: 0.9004387714399682
Overall Recall: 0.8869565217391304
Overall F1 Score: 0.8611495227995759


> **Model Evaluation Summary**

>The model achieves an overall accuracy of ~88.7%, but performance varies significantly across classes.

>For **Class 0 (Normal)**, results are good: high precision (0.88), recall (1), and F1 (0.93) show the model is very reliable at identifying non‑diabetic patients.

>For **Class 1 (Diabetic)**, performance is moderate: precision (1) is good, but recall (0.31) indicates many diabetic cases are missed, leading to a weaker F1 (0.48).

>Overall precision (0.9), recall (0.87), and F1 (0.86) looks good, but they are heavily skewed by the dominance of Class 0. The imbalance in the dataset causes the model to favor the majority class while neglecting minority class. To improve Class 1 performance, techniques such as oversampling (SMOTE) can be applied to give more importance to the minority class during training.